doc/source/indexing.rst

```diff
@@ -667,6 +667,94 @@ For getting *multiple* indexers, using ``.get_indexer``
   dfd.iloc[[0, 2], dfd.columns.get_indexer(['A', 'B'])]
 
 
+.. _indexing.deprecate_loc_reindex_listlike:
+
+Using loc with missing keys in a list is Deprecated
+---------------------------------------------------
+
+.. warning::
+
+   Starting in 0.21.0, using ``.loc`` with a list-like containing missing key, is deprecated, in favor of ``.reindex``.
+
+In prior versions, using ``.loc[list-of-keys]`` would work as long as *at least 1* of the keys was found (otherwise it
+would raise a ``KeyError``). This behavior is deprecated and will show a warning message pointing to this section. The
+recommeded alternative is to use ``.reindex()``.
+
+For example.
+
+.. ipython:: python
+
+   s = Series([1, 2, 3])
+   s
+
+Selection with all keys found is unchanged.
+
+.. ipython:: python
+
+   s.loc[[1, 2]]
+
+Previous Behavior
+
+.. code-block:: ipython
+
+
+   In [4]: s.loc[[1, 2, 3]]
+   Out[4]:
+   1    2.0
+   2    3.0
+   3    NaN
+   dtype: float64
+
+
+Current Behavior
+
+   In [4]: s.loc[[1, 2, 3]]
+   Passing list-likes to .loc with any non-matching elements will raise
+   KeyError in the future, you can use .reindex() as an alternative.
+
+   See the documentation here:
+   http://pandas.pydata.org/pandas-docs/stable/indexing.html#deprecate-loc-reindex-listlike
+
+   Out[4]:
+   1    2.0
+   2    3.0
+   3    NaN
+   dtype: float64
+
+
+Reindexing
+~~~~~~~~~~
+
+The idiomatic way to achieve selecting potentially not-found elmenents is via ``.reindex()``. See also the section on :ref:`reindexing <basics.reindexing>`.
+
+.. ipython:: python
+
+  s.reindex([1, 2, 3])
+
+Alternatively, if you want to select only *valid* keys, the following is idiomatic; furthermore this is more efficient, and is guaranteed to preserve the dtype of the selection.
+
+.. ipython:: python
+
+   keys = [1, 2, 3]
+   s.loc[s.index & keys]
+
+Having a duplicated index will raise for a ``.reindex()``:
+
+.. ipython:: python
+
+   s = pd.Series(np.arange(4), index=['a', 'a', 'b', 'c'])
+
+.. code-block:: python
+
+   In [17]: s.reindex(['c', 'd'])
+   ValueError: cannot reindex from a duplicate axis
+
+The idiomatic expression again allows this operation to proceed
+
+.. ipython:: python
+
+   s.loc[s.index & ['c', 'd']]
```

Comment:
> Is there actually a way to get exactly the same?

Not a trivial one. My point is not that with the intersection it is a drop-in replacement, just that it _should_ be, in the right way to proceed. Anyway, the best answer to your question is, I think: people who were ``.loc``ing on purpose for non-existing keys in a unique index can use ``.reindex()``; people who instead where ``.loc``ing for non-existing keys in a non-unique index were doing it just wrong (because the resulting ``NaN``s have no meaning in a non-unique index), and should rather use the intersection.